Fix: Production docker image built without test gems

dragon-dxw opened the pull request:
Backport of https://github.com/citizensadvice/corporate-api/pull/55

The CA Corporate API is headless, so it doesn't use node, which means it's likely that this doesn't yet correctly handle node modules.

## Changes in this PR

Before this change our Dockerfile wasn't correctly creating Docker images. It was bundling gems from the `test` group despite expecting `RAILS_ENV` to be set as `production` by default. We don't want to be deploying live artefacts that have test dependencies and tooling installed. They are bigger and increase the surface area for complexity and attacks.

The meat of this change is centered around the Dockerfile. Instead of using `RAILS_ENV` to try and let the user convey what gems they expected, we switch to creating more explicit docker build stages and select them with the `target` field instead.

## Testing

1. `DOCKER=1 script/test` should still run all the tests (and have the test gems to do so)
2. `DOCKER=1 script/server` should still start up a local development server and be accesible at localhost:3000

If you want to test this in production too you can temporarily copy this into your app directory, into a file called `docker-compose.production.yml`. Once there you can run `docker-compose -f docker-compose.production.yml up` to check it boots in the right environment and watch that the right gems were installed: 

```
version: "3.8"
services:
  web:
    build:
      context: .
      target: production
    command: bash -c "rm -f tmp/pids/server.pid && bundle exec rails s -p 3000 -b '0.0.0.0'"
    ports:
      - "3000:3000"
    depends_on:
      - db
    env_file:
       - .env.development.local
    environment:
      # DONE: Update rails-template with application name
      DATABASE_URL: postgres://postgres:password@db:5432/corporate-api-production
      AUTOMATICALLY_FIX_LINTING: "true"
    volumes:
      - .:/srv/app
    tty: true
    stdin_open: true
    networks:
      - prod

  db:
    image: postgres
    volumes:
      - db-data:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: password
      POSTGRES_HOST_AUTH_METHOD: trust
    networks:
      - prod

networks:
  prod:

volumes:
  db-data:
```

tahb commented:
These changes worked well for CAEW and we should get this merged in!

pezholio commented:
I'd like to get this in, but the tests are failing at the mo

tahb commented:
I've checked with Dragon and I'm going to try and move this along.

tahb committed:
By default the Dockerfile only installs production gems

`bundle config set with $BUNDLE_GEM_GROUPS` wasn't doing what we thought in the production context.

It'd work out to be `bundle config set with production` which would bundle everything, rather than only production groups.
I'm not sure where this assumption came from. 'default' also doesn't work.

As the first step of a bit of a refactoring of the Dockerfile, we cement the assurance that by default the dockerfile _only_ bundles production gems.
We remove the way the build_arg `RAILS_ENV` was used to accomodate being built for production, development and test. This hopefully makes it much clearer.

(I've also set NODE_ENV to follow the same path as RAILS_ENV -- Dragon)

tahb committed:
Dockerfile installs development and test gems in unique build stage

Instead of passing `RAILS_ENV` as a build argument to the dockerfile to figure out what should be bundled.
We switch to using `--target` to target a specific docker build stage that hardcodes the `RAILS_ENV` so it can't have an unexpected value.

We now have a stage for production, development and test. Each explicitly installs gems for the right group.

tahb committed:
Fix JS compression error by running Ugilifier in harmony mode

When building the docker image in production with script/docker/update
we get a new error:

```
 > [production 21/23] RUN   if [ "production" = "production" ]; then   SECRET_KEY_BASE="secret"   bundle exec rake assets:precompile;   fi:
43 3.045 Yarn executable was not detected in the system.
43 3.045 Download Yarn at https://yarnpkg.com/en/docs/install
43 3.326 Yarn executable was not detected in the system.
43 3.326 Download Yarn at https://yarnpkg.com/en/docs/install
43 3.668 rake aborted!
43 3.668 Uglifier::Error: Unexpected token: keyword (const). To use ES6 syntax, harmony mode must be enabled with Uglifier.new(:harmony => true).
…
43 3.668  11663   /**
43 3.668  11664    * ------------------------------------------------------------------------
43 3.668  11665    * Constants
43 3.668  11666    * ------------------------------------------------------------------------
43 3.668  11667    */
43 3.668     =>   const elementMap = new Map();
```

We follow the instruction to use `Uglifier.new(:harmony => true)` and
sure enought the error is silenced.

tahb commented:
@pezholio back in business with a green test suite 👍

erbridge commented on Rakefile:
```diff
@@ -1,11 +1,11 @@
-# frozen_string_literal: true
-
-# Add your own tasks in files placed in lib/tasks ending in .rake,
```
This line went missing somehow.

erbridge commented on Rakefile:
```diff
@@ -1,11 +1,11 @@
 
-desc "Run all the tests"
-task default: %i[spec standard]
+  RuboCop::RakeTask.new
+end
+
+Rails.application.load_tasks
```
This means we no longer run the tests and standard as the default rake task. Is that intentional?

tahb replied:
Nope! Dropping this commit and starting again.

erbridge commented on Rakefile:
```diff
@@ -1,11 +1,11 @@
-Rails.application.load_tasks
+if Rails.env.development? || Rails.env.test?
+  require "rubocop/rake_task"
 
-desc "Run all the tests"
-task default: %i[spec standard]
+  RuboCop::RakeTask.new
+end
```
Rubocop and not Standard? Is that right?

tahb replied:
Yes it should be Standard, this change ports a change from a repo that used Rubocop. I'm going to drop this commit and make a new one.

erbridge commented on Dockerfile:
```diff
@@ -118,10 +112,32 @@ EXPOSE 3000
+RUN bundle config set with development
+RUN bundle install --retry=10 --jobs=4
+
+# Define the runtime command in docker-compose.yml
```
Can we set a placeholder command like rails console or something so it will do something different from the production image even if we don't remember to set the command in the Docker Compose config.

tahb replied:
Sure

erbridge commented on Dockerfile:
```diff
@@ -118,10 +112,32 @@ EXPOSE 3000
+ENV RAILS_ENV=test
+ENV NODE_ENV=test
+
+RUN bundle config unset without
```
We're inheriting from the `development` image here, so we don't need to repeat this. Though I'm not sure why we're inheriting from `development` and not `production`?

tahb replied:
Yes I think this should be inheriting from production so I'll change that and leave `RUN bundle config unset without` as it was.

erbridge commented on Dockerfile:
```diff
@@ -136,3 +152,7 @@ COPY .prettierrc ${APP_HOME}/.prettierrc
 COPY .rspec ${APP_HOME}/.rspec
 COPY spec ${APP_HOME}/spec
+
+RUN yarn install
```
Is this intentionally not freezing the lockfile?

tahb replied:
For CI a locked dependency file makes sense. Is there the case during local testing where we might want to add or change a package without rebuilding the whole image? Perhaps it's rare enough that this is ok. I'll lock it for now.

erbridge commented on Dockerfile:
```diff
@@ -136,3 +152,7 @@ COPY .prettierrc ${APP_HOME}/.prettierrc
+
+RUN yarn install
+
+# Define the runtime command in docker-compose.test.yml
```
Can we set a placeholder command like rails console or something so it will do something different from the production image even if we don't remember to set the command in the Docker Compose config.

tahb replied:
Sure. I have put `rake` as that should run the tests and be a sensible default.

erbridge commented on config/environments/production.rb:
```diff
@@ -27,7 +27,7 @@
 
   # Compress JavaScripts and CSS.
-  config.assets.js_compressor = :uglifier
+  config.assets.js_compressor = Uglifier.new(harmony: true)
```
Can we explain what this is doing in the commit message, please?

erbridge replied:
Commit message is missing the error message @tahb. I assume you meant to copy it in?

tahb replied:
Yes I did mean to copy it in, fail.

Added it in now and checked:

```
 > [production 21/23] RUN   if [ "production" = "production" ]; then   SECRET_KEY_BASE="secret"   bundle exec rake assets:precompile;   fi:
43 3.045 Yarn executable was not detected in the system.
43 3.045 Download Yarn at https://yarnpkg.com/en/docs/install
43 3.326 Yarn executable was not detected in the system.
43 3.326 Download Yarn at https://yarnpkg.com/en/docs/install
43 3.668 rake aborted!
43 3.668 Uglifier::Error: Unexpected token: keyword (const). To use ES6 syntax, harmony mode must be enabled with Uglifier.new(:harmony => true).
…
43 3.668  11663   /**
43 3.668  11664    * ------------------------------------------------------------------------
43 3.668  11665    * Constants
43 3.668  11666    * ------------------------------------------------------------------------
43 3.668  11667    */
43 3.668     =>   const elementMap = new Map();
```

tahb committed:
Application starts in production mode

When starting the application with RAILS_ENV=production mode and we run `rake` we get an error that explains we don't have standard installed.

This is correct as out gemfile specifies that we only install standard in the dev and test environments.

This change only tries to load the standard rake task when in the dev or test environment.

Other take tasks such as `rails db:prepare` can still run in any environment as normal.

tahb commented:
@erbridge I completely forgot about this! I've actioned your feedback now.

erbridge reviewed (approved):
Looks sensible to me. Minor commit message issue, but otherwise 👍.

dragon-dxw commented:
LGTM, I can't approve as the creator, but @tabh feel free to approve on my behalf if you're happy with it :)

tahb reviewed (approved):
Approving on behalf of Dragon who's the original author and therefore can't :)

lozette commented:
Been trying to get the tests passing on this to no avail, sorry!

dragon-dxw commented:
Converted to draft because this is no longer fit to be published, but someone might want to build on this work in future.

Also it'll silence the notifications I get for it...

mec commented:
Given the state and age of this work, I am closing this PR.
